### DISCUSSION TITLE
Numbered dimmers do not work as triggers in rules

### DISCUSSION THREAD

**Discussion by mvn23** [arendst/Tasmota#16873] [Q&A] [answered]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._
When trying to respond to dimmer state changes, only dimmer1 can be used as a trigger without specifying a dimmer number (e.g. `DIMMER#State=100`).
I have recently started using some Shelly RGBW2 LED dimmers. I have `SetOption37` set to `ON` to separate the RGB control from the White control. This gives me 2 dimmers: `dimmer1` for the RGB channel and `dimmer2` for the White channel. I can use dimmer1 as a rule trigger, but only without specifying the number in the trigger, i.e. `DIMMER#State=100` works, but `DIMMER1#State=100` does not.
Furthermore, `DIMMER2#State=100` does not work either. This makes it impossible to respond to dimmer state changes for dimmers > 1.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): Shelly RGBW2
- [x] Tasmota binary firmware version number used: 12.2.0
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: esptool
- [] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
09:38:32.127 CMD: Backlog Template; Module; GPIO 255
09:38:32.165 MQT: stat/light-dinnertable/RESULT = {"NAME":"ShellyRGBW2","GPIO":[0,0,544,0,419,193,0,0,416,32,418,417,0,4896],"FLAG":0,"BASE":18}
09:38:32.378 MQT: stat/light-dinnertable/RESULT = {"Module":{"0":"ShellyRGBW2"}}
09:38:32.593 MQT: stat/light-dinnertable/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"544":"LedLink"},"GPIO3":{"0":"None"},"GPIO4":{"419":"PWM4"},"GPIO5":{"193":"Switch_n2"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"416":"PWM1"},"GPIO13":{"32":"Button1"},"GPIO14":{"418":"PWM3"},"GPIO15":{"417":"PWM2"},"GPIO16":{"0":"None"},"GPIO17":{"4896":"ADC CT Power"}}
```
- [x] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
09:42:35.125 MQT: stat/light-dinnertable/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":47,"Free":464,"Rules":"ON dimmer1#state=100 DO backlog mem1 test ENDON"}}
09:42:35.346 MQT: stat/light-dinnertable/RESULT = {"Rule2":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
09:42:35.595 MQT: stat/light-dinnertable/RESULT = {"Rule3":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
09:45:20.418 MQT: stat/light-dinnertable/STATUS = {"Status":{"Module":0,"DeviceName":"light-dinnertable","FriendlyName":["rgb-dinnertable","light-dinnertable"],"Topic":"light-dinnertable","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[11,11,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
09:45:20.433 MQT: stat/light-dinnertable/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Power On","Uptime":"0T19:08:07","StartupUTC":"2022-10-20T13:37:13","Sleep":50,"CfgHolder":4617,"BootCount":18,"BCResetTime":"2022-10-20T12:01:50","SaveCount":501,"SaveAddress":"F7000"}}
09:45:20.443 MQT: stat/light-dinnertable/STATUS2 = {"StatusFWR":{"Version":"12.2.0(tasmota)","BuildDateTime":"2022-10-17T08:39:23","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"432/699"}}
09:45:20.455 MQT: stat/light-dinnertable/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["SkyNet",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["0000800B","0A05C80001800600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
09:45:20.471 MQT: stat/light-dinnertable/STATUS4 = {"StatusMEM":{"ProgramSize":626,"Free":376,"Heap":24,"ProgramFlashSize":2048,"FlashSize":2048,"FlashChipId":"156085","FlashFrequency":40,"FlashMode":"DOUT","Features":["00000809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","54000020","00000000"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62","Sensors":"1,2,3,4,5,6"}}
09:45:20.483 MQT: stat/light-dinnertable/STATUS5 = {"StatusNET":{"Hostname":"light-dinnertable-1214","IPAddress":"10.0.22.150","Gateway":"10.0.22.1","Subnetmask":"255.255.255.0","DNSServer1":"10.0.22.254","DNSServer2":"0.0.0.0","Mac":"58:BF:25:D9:24:BE","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
09:45:20.494 MQT: stat/light-dinnertable/STATUS6 = {"StatusMQT":{"MqttHost":"10.0.22.254","MqttPort":1883,"MqttClientMask":"light-dinnertable","MqttClient":"light-dinnertable","MqttUser":"hass","MqttCount":5,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
09:45:20.506 MQT: stat/light-dinnertable/STATUS7 = {"StatusTIM":{"UTC":"2022-10-21T08:45:20","Local":"2022-10-21T09:45:20","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":"+01:00","Sunrise":"07:19","Sunset":"17:49"}}
09:45:20.559 MQT: stat/light-dinnertable/STATUS10 = {"StatusSNS":{"Time":"2022-10-21T09:45:20","Switch2":"OFF","ANALOG":{"CTEnergy":{"Energy":0.009,"Power":0,"Voltage":12,"Current":0.000}}}}
09:45:20.574 MQT: stat/light-dinnertable/STATUS11 = {"StatusSTS":{"Time":"2022-10-21T09:45:20","Uptime":"0T19:08:07","UptimeSec":68887,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":26,"MqttCount":5,"POWER1":"OFF","Dimmer1":0,"POWER2":"OFF","Dimmer2":37,"Color":"0000005F","HSBColor":"0,0,0","White":37,"Channel":[0,0,0,37],"Scheme":0,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"SkyNet","BSSId":"76:83:C2:2D:A7:DB","Channel":6,"Mode":"11n","RSSI":82,"Signal":-59,"LinkCount":2,"Downtime":"0T00:00:09"}}}
```
- [x] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
09:49:20.354 CMD: dimmer1 100
09:49:20.356 SRC: WebConsole from 192.168.23.23
09:49:20.359 CMD: Grp 0, Cmd 'DIMMER', Idx 1, Len 3, Pld 100, Data '100'
09:49:20.363 SRC: Light
09:49:20.374 MQT: stat/light-dinnertable/RESULT = {"POWER1":"ON","Dimmer1":100,"POWER2":"OFF","Dimmer2":37,"Color":"FFFFFF5F","HSBColor":"0,0,100","White":37,"Channel":[100,100,100,37]}
09:49:20.956 CFG: Saved to flash at F8, Count 508, Bytes 4096
```

### TO REPRODUCE
_Steps to reproduce the behavior:_
* Create a rule with a numbered dimmer as trigger, e.g. `DIMMER1#State=100`
* Try to trigger the rule

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
The rule should be usable.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by sfromis** [comment, marked as answer]
No bug here, but you may like an enhancement to the built-in triggers.

One workaround is to use JSON-based triggers instead of the single built-in dimmer trigger.
```
on Dimmer1 do var1 %value% endon
on Dimmer2 do var2 %value% endon
```
These will also fire when STATE payloads are published, and with two triggers in the same payload, both will fire at the same time. Thus they do not really indicate that dimmer state was changed.

Such JSON-based triggers cannot work the same way if there are built-in triggers for the same source.

**Comment by mvn23** [comment]
Just now I noticed that it does indeed work as described in the [docs](https://tasmota.github.io/docs/Rules/#examples-of-available-triggers) (although the list there just states `examples`, suggesting it is not exhaustive), so it is indeed not a bug. However, since we have similar `POWER1#state`, `BUTTON1#state`, etc. triggers, would it not make sense to apply the same logic to dimmers?

I can't comment on the technical possibility of JSON vs. native triggers as I have insufficient knowledge of the inner workings involved, but to me it would make sense to support a `DIMMER1#state` trigger just like all other numbered triggers.

Edit: for now I'm using the workaround you suggested together with a `VAR1#state` trigger, but it does not seem to work as smooth as the native `DIMMER#state` logic I use on other devices. That could be some other setting throwing a spanner in the works though...

**Reply by sfromis** [reply]
The point of "examples" is that there can be lots of JSON-based triggers not listed there, but for builtin triggers the list should be "structurally exhaustive", meaning that without `<x>` or some number, nothing such is meant. Sure, it would "make sense" to have what you ask for, but the set of what would "make sense" is vast.

While I'd certainly not be against it, if someone decides to work on such an enhancement, there are lots of thing around support for lights which are squarely designed around only having one light available, and many other changes would make as much sense to fully support more than one.
